src/helpers/getByAPI.js

```diff
@@ -173,13 +229,28 @@ export const getAllByProps = (instance: ReactTestInstance) =>
     return results;
   };
 
+export const getAllByTestId = (instance: ReactTestInstance) =>
+  function getAllByTestIdFn(testID: string) {
+    const results = instance.findAllByProps({ testID });
```

Comment:
As you noticed in the OP, because `findAllByProps` returns all components with such ID (so every if the component have a `testID` prop that is passed further and further, more than one component will be matched).
We want to filter the results to only match the `element`s which `element.type` is a string – this is the way we know the renderer rendered the "host node" (in case of RN, it's usually `View` or `Text`).

Something like this should be fine:
```js
const results = instance.findAllByProps({ testID })
  .filter(element => typeof element.type === 'string')
```

FYI, we're not currently doing this filtering for `getAllByProps` et. al. queries, which is a bug. But we're removing these in v2 anyway.

Reply:
Yeah that's a terrible solution, it won't work if you put `testID` on a custom component. My solution doesn't have that issue.

Reply:
By the way, this code: 
```
const { getAllByTestId } = render(
  <TouchableOpacity testID="testID" activeOpacity={0.5} />
)
console.log(getAllByTestId("testID").map(({ props, type }) => ({ props, type })))
```
Shows this: 
```
      [ { props: { testID: 'testID', activeOpacity: 0.5 },
          type:
           { [Function]
             displayName: 'TouchableOpacity',
             propTypes: [Object],
             getDefaultProps: [Function],
             defaultProps: [Object] } },
        { props:
           { accessible: true,
             accessibilityLabel: undefined,
             accessibilityHint: undefined,
             accessibilityRole: undefined,
             accessibilityStates: undefined,
             style: [Array],
             nativeID: undefined,
             testID: 'testID',
             onLayout: undefined,
             isTVSelectable: true,
             hasTVPreferredFocus: undefined,
             tvParallaxProperties: undefined,
             hitSlop: undefined,
             onStartShouldSetResponder: [Function],
             onResponderTerminationRequest: [Function],
             onResponderGrant: [Function],
             onResponderMove: [Function],
             onResponderRelease: [Function],
             onResponderTerminate: [Function],
             children: [Array] },
          type:
           { [Function: AnimatedComponent]
             __skipSetNativeProps_FOR_TESTS_ONLY: false,
             propTypes: [Object] } },
        { props:
           { accessible: true,
             accessibilityLabel: undefined,
             accessibilityHint: undefined,
             accessibilityRole: undefined,
             accessibilityStates: undefined,
             style: [Object],
             nativeID: undefined,
             testID: 'testID',
             onLayout: undefined,
             isTVSelectable: true,
             hasTVPreferredFocus: undefined,
             tvParallaxProperties: undefined,
             hitSlop: undefined,
             onStartShouldSetResponder: [Function],
             onResponderTerminationRequest: [Function],
             onResponderGrant: [Function],
             onResponderMove: [Function],
             onResponderRelease: [Function],
             onResponderTerminate: [Function],
             children: [Array],
             collapsable: undefined },
          type:
           { [Function: Component]
             displayName: 'View',
             '$$typeof': Symbol(react.forward_ref),
             render: [Function: View] } },
        { props:
           { accessible: true,
             style: [Object],
             testID: 'testID',
             isTVSelectable: true,
             onStartShouldSetResponder: [Function],
             onResponderTerminationRequest: [Function],
             onResponderGrant: [Function],
             onResponderMove: [Function],
             onResponderRelease: [Function],
             onResponderTerminate: [Function],
             children: [Array] },
          type: 'View' } ]
```
I don't think I need to explain why choosing the last element as the only one that the function should return is a terrible idea.

Reply:
The idea behind `testID` is to be applied to the element that directly translates to the native View (or other). If your component is implemented properly and passes props down to children, React should pass them to this very host element ("native" View). Firing events will also work, since they bubble up in the component hierarchy.

If you think it's terrible, feel free to close this PR and create your own fork of the library. I cannot simply accept hacks like generic `getBy` and `getByAll` helpers because the point of this library is to be opinionated and prevent you from accessing props wherever possible (that's why we e.g. remove `getByProps` helpers in v2).